### PR TITLE
Hive: New test case for writing to the table with projection and column specification

### PR DESCRIPTION
**Pull request by pvary**
During the test found several issues with vectorization for which I raised #2120 

There are no issues with vectorization turned off, but having a test case for it helps finding possible issues later

### CODE REVIEW

**Comment by pvary** [comment]
@marton-bod, @lcspinter: Could you please review?

Thanks,
Peter

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 1]
```diff
@@ -335,6 +335,26 @@ public void testInsertFromSelectWithOrderBy() throws IOException {
...
 
+  @Test
+  public void testInsertFromSelectWithProjection() throws IOException {
+    Assume.assumeTrue("Tez write is not implemented yet", executionEngine.equals("mr"));
+
+    Table table = testTables.createTable(shell, "customers", HiveIcebergStorageHandlerTestUtils.CUSTOMER_SCHEMA,
+        fileFormat, ImmutableList.of());
+    testTables.createTable(shell, "orders", ORDER_SCHEMA, fileFormat, ORDER_RECORDS);
+
+    shell.setHiveSessionValue("hive.vectorized.execution.enabled", false);
```
Comment (lcspinter) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:346]: Vectorization was already disabled in TestHiveShell.initializeConf? Do we need to disable it again?

**Review by marton-bod — APPROVED** [review]
LGTM, my only question would be the same as Laszlo's

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 1 above)
Comment (pvary) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java]: Yeah, I have missed this.
Thanks for catching!

**Review by lcspinter — APPROVED** [review]
+1 (non-binding)

**Comment by pvary** [comment]
@rdblue: Could you please take a look at this PR? Only adds a single test case which failed for tez writes. Nothing fancy, but good to have a test case for MR as well.

Thanks,
Peter

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveShell.java [hunk 2]
```diff
@@ -186,6 +186,9 @@ private HiveConf initializeConf() {
     // Tez configuration
     hiveConf.setBoolean("tez.local.mode", true);
 
+    // Disable vectorization for HiveIcebergInputFormat
+    hiveConf.setBoolVar(HiveConf.ConfVars.HIVE_VECTORIZATION_ENABLED, false);
```
Comment (rdblue) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveShell.java:190]: I think this is fine, but I'm curious: do we plan to add vectorized writes? Or is this not something that we think is important?

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 3]
```diff
@@ -335,6 +335,25 @@ public void testInsertFromSelectWithOrderBy() throws IOException {
...
+
+    shell.executeStatement(
+        "INSERT INTO customers (customer_id, last_name) SELECT distinct(customer_id), 'test' FROM orders");
+
+    List<Record> expected = TestHelper.RecordsBuilder.newInstance(HiveIcebergStorageHandlerTestUtils.CUSTOMER_SCHEMA)
+        .add(0L, null, "test")
+        .add(1L, null, "test")
+        .build();
+
+    HiveIcebergTestUtils.validateData(table, new ArrayList<>(expected), 0);
```
Comment (rdblue) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:354]: Why does this create a new `ArrayList` from the expected records?

**Comment by rdblue** [comment]
Overall, this looks fine. One minor question about why it creates an extra list. Otherwise +1.

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 3 above)
Comment (pvary) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:354]: `HiveIcebergTestUtils.validateData` uses in place sort to compare the actual and the expected records. This is needed of other uses of the method, so I just recreated a new list for the test case

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveShell.java (hunk 2 above)
Comment (pvary) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveShell.java:190]: I think it is important, but no hard data on yet. Will work on read/write performance in the near future, so  would guess that vectorization will be something we check too.

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 3 above)
Comment (rdblue) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:354]: Doesn't this leak the concern about sorting outside of the method? I would rather have the array copy happen defensively inside the method.

Comment (pvary) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java]: Good point. Fixed here, and all the relevant other places

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @pvary!

**Comment by pvary** [comment]
Thanks for the review and the merge @rdblue!

**Pull request merged**